File: framework/languages/java/e2e/PubSub/Server/Subscriber/src/main/java/systems/zlink/e2e/pubsub/subscriber/Endpoints/OperationalEndpoints.java
package Endpoints;

import com.sun.net.httpserver.HttpExchange;
import systems.zlink.e2e.pubsub.subscriber.Configuration;
import systems.zlink.e2e.pubsub.subscriber.Endpoints;
import systems.zlink.e2e.pubsub.subscriber.Infrastructure;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import org.springframework.context.SmartLifecycle;
import systems.zlink.e2e.pubsub.subscriber.Configuration.SubscriberOptions;
import systems.zlink.e2e.pubsub.subscriber.Infrastructure.EvidenceStore;
import systems.zlink.e2e.pubsub.subscriber.Infrastructure.FanoutObserverController;
import systems.zlink.e2e.pubsub.subscriber.Infrastructure.SubscriberConnections;
import systems.zlink.e2e.pubsub.shared.Contracts;
import systems.zlink.framework.monitoring.ZLinkFanoutRuntime;
import systems.zlink.framework.monitoring.ZLinkPeerState;

public final class OperationalEndpoints implements SmartLifecycle {
    private final SubscriberOptions options;
    private final EvidenceStore evidence;
    private final ObjectMapper json;
    private final ZLinkFanoutRuntime fanoutRuntime;
    private final SubscriberConnections connections;
    private final FanoutObserverController observers;
    private HttpServer server;
    private boolean running;

    public OperationalEndpoints(
        SubscriberOptions options,
        EvidenceStore evidence,
        ObjectMapper json,
        ZLinkFanoutRuntime fanoutRuntime,
        SubscriberConnections connections,
        FanoutObserverController observers) {
        this.options = options;
        this.evidence = evidence;
        this.json = json;
        this.fanoutRuntime = fanoutRuntime;
        this.connections = connections;
        this.observers = observers;
    }

    @Override
    public void start() {
        if (options.httpEndpoint() == null || options.httpEndpoint().isBlank()) {
            return;
        }
        try {
            URI uri = URI.create(options.httpEndpoint());
            server = HttpServer.create(new InetSocketAddress(uri.getHost(), uri.getPort()), 0);
            server.createContext("/health", exchange -> {
                byte[] body = "ok\n".getBytes(StandardCharsets.UTF_8);
                exchange.sendResponseHeaders(200, body.length);
                exchange.getResponseBody().write(body);
                exchange.close();
            });
            server.createContext("/evidence", exchange -> {
                byte[] body = json.writeValueAsBytes(evidence.snapshot());
                exchange.getResponseHeaders().add("Content-Type", "application/json");
                exchange.sendResponseHeaders(200, body.length);
                exchange.getResponseBody().write(body);
                exchange.close();
            });
            // Config 3의 publisher 상태 전환을 public monitoring snapshot으로
            // 관측한다. 내부 Location Store의 descriptor를 process 밖으로 노출하지 않는다.
            server.createContext("/locations/publishers", exchange -> {
                var publishers = fanoutRuntime.snapshot(Contracts.EVENT_CHANNEL).publishers().stream()
                    .filter(publisher -> publisher.state() == ZLinkPeerState.READY)
                    .map(publisher -> publisher.nodeRid().toString())
                    .toList();
                byte[] body = json.writeValueAsBytes(publishers);
                exchange.getResponseHeaders().add("Content-Type", "application/json");
                exchange.sendResponseHeaders(200, body.length);
                exchange.getResponseBody().write(body);
                exchange.close();
            });
            server.createContext("/status", exchange -> {
                var status = fanoutRuntime.snapshot(Contracts.EVENT_CHANNEL);
                writeJson(exchange, Map.of(
                    "channelName", status.channelName(),
                    "state", status.state().name(),
                    "isReady", status.isReady(),
                    "readyPublisherCount", status.readyPublisherCount(),
                    "publishers", status.publishers().stream().map(publisher -> Map.of(
                        "nodeRid", publisher.nodeRid().toString(),
                        "state", publisher.state().name())).toList(),
                    "sequence", status.sequence(),
                    "observedAt", status.observedAt().toString()));
            });
            server.createContext("/connections", exchange -> {
                try {
                    Map<String, String> values = query(exchange.getRequestURI());
                    switch (values.getOrDefault("operation", "list")) {
                        case "connect" -> connections.connect(required(values, "endpoint"));
                        case "disconnect" -> connections.disconnect(required(values, "endpoint"));
                        case "list" -> { }
                        default -> throw new IllegalArgumentException("unsupported connection operation");
                    }
                    writeJson(exchange, Map.of("connections", connections.list()));
                } catch (Exception error) {
                    writeText(exchange, 400, error.getMessage() + "\n");
                }
            });
            server.createContext("/observer/start", exchange -> {
                try {
                    Map<String, String> values = query(exchange.getRequestURI());
                    observers.start(
                        values.getOrDefault("name", "normal"),
                        intQuery(values, "capacity", 1),
                        Boolean.parseBoolean(values.getOrDefault("slow", "false")));
                    writeJson(exchange, Map.of("status", "started"));
                } catch (Exception error) {
                    writeText(exchange, 400, error.getMessage() + "\n");
                }
            });
            server.createContext("/observer/release", exchange -> {
                observers.release(query(exchange.getRequestURI()).getOrDefault("name", "slow"));
                writeJson(exchange, Map.of("status", "released"));
            });
            server.createContext("/observer/cancel", exchange -> {
                observers.cancel(query(exchange.getRequestURI()).getOrDefault("name", "slow"));
                writeJson(exchange, Map.of("status", "cancelled"));
            });
            server.createContext("/observer/wait", exchange -> {
                try {
                    Map<String, String> values = query(exchange.getRequestURI());
                    observers.waitFor(
                        values.getOrDefault("name", "normal"),
                        longQuery(values, "timeoutMs", 30_000L));
                    writeJson(exchange, Map.of("status", "observed"));
                } catch (Exception error) {
                    writeText(exchange, 504, error.getMessage() + "\n");
                }
            });
            server.createContext("/observer/evidence", exchange ->
                writeJson(exchange, observers.snapshot()));
            server.createContext("/evidence/wait", exchange -> {
                try {
                    Map<String, String> query = query(exchange.getRequestURI());
                    long timeoutMillis = longQuery(query, "timeoutMillis", 15_000L);
                    var snapshot = switch (query.getOrDefault("kind", "event")) {
                        case "any-event" -> evidence.waitFor(
                            entry -> "EventMsg".equals(entry.marker())
                                && query.getOrDefault("scenario", "").equals(entry.scenario()),
                            timeoutMillis);
                        case "event" -> evidence.waitFor(
                            entry -> "EventMsg".equals(entry.marker())
                                && query.getOrDefault("scenario", "").equals(entry.scenario())
                                && entry.sequence() == intQuery(query, "sequence", Integer.MIN_VALUE),
                            timeoutMillis);
                        case "sequence-at-least" -> evidence.waitFor(
                            entry -> "EventMsg".equals(entry.marker())
                                && query.getOrDefault("scenario", "").equals(entry.scenario())
                                && entry.sequence() >= intQuery(query, "minSequence", Integer.MAX_VALUE),
                            timeoutMillis);
                        case "dispatch-error" -> evidence.waitFor(
                            entry -> "DispatchError".equals(entry.marker())
                                && entry.value().contains("HANDLER_MISSING")
                                && entry.value().contains("DROP")
                                && entry.value().contains(query.getOrDefault("packetName", "")),
                            timeoutMillis);
                        case "contiguous-run" -> evidence.waitForContiguousRun(
                            query.getOrDefault("scenario", ""),
                            intQuery(query, "minLength", 1),
                            timeoutMillis);
                        default -> throw new IllegalArgumentException("unknown evidence wait kind");
                    };
                    byte[] body = json.writeValueAsBytes(snapshot);
                    exchange.getResponseHeaders().add("Content-Type", "application/json");
                    exchange.sendResponseHeaders(200, body.length);
                    exchange.getResponseBody().write(body);
                } catch (Exception error) {
                    byte[] body = (error.getMessage() + "\n").getBytes(StandardCharsets.UTF_8);
                    exchange.sendResponseHeaders(408, body.length);
                    exchange.getResponseBody().write(body);
                } finally {
                    exchange.close();
                }
            });
            server.start();
            running = true;
        } catch (Exception error) {
            throw new IllegalStateException("failed to start evidence endpoint " + options.httpEndpoint(), error);
        }
    }

    @Override
    public void stop() {
        if (server != null) {
            server.stop(0);
            server = null;
        }
        running = false;
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    private static Map<String, String> query(URI uri) {
        Map<String, String> result = new HashMap<>();
        String raw = uri.getRawQuery();
        if (raw == null || raw.isBlank()) {
            return result;
        }
        for (String pair : raw.split("&")) {
            int separator = pair.indexOf('=');
            if (separator <= 0) {
                continue;
            }
            String key = decode(pair.substring(0, separator));
            String value = decode(pair.substring(separator + 1));
            result.put(key, value);
        }
        return result;
    }

    private static String decode(String value) {
        return URLDecoder.decode(value, StandardCharsets.UTF_8);
    }

    private static int intQuery(Map<String, String> query, String name, int fallback) {
        String value = query.get(name);
        return value == null || value.isBlank() ? fallback : Integer.parseInt(value);
    }

    private static long longQuery(Map<String, String> query, String name, long fallback) {
        String value = query.get(name);
        return value == null || value.isBlank() ? fallback : Long.parseLong(value);
    }

    private static String required(Map<String, String> values, String name) {
        String value = values.get(name);
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(name + " is required");
        }
        return value;
    }

    private void writeJson(HttpExchange exchange, Object value) {
        try {
            byte[] body = json.writeValueAsBytes(value);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, body.length);
            exchange.getResponseBody().write(body);
            exchange.close();
        } catch (Exception error) {
            throw new IllegalStateException("failed to write subscriber json response", error);
        }
    }

    private static void writeText(
        HttpExchange exchange,
        int status,
        String value) {
        try {
            byte[] body = value.getBytes(StandardCharsets.UTF_8);
            exchange.sendResponseHeaders(status, body.length);
            exchange.getResponseBody().write(body);
            exchange.close();
        } catch (Exception error) {
            throw new IllegalStateException("failed to write subscriber response", error);
        }
    }
}

File: framework/languages/java/e2e-kotlin/PubSub/Server/Subscriber/src/main/kotlin/systems/zlink/e2e/kotlin/pubsub/subscriber/Endpoints/OperationalEndpoints.kt
package systems.zlink.e2e.kotlin.pubsub.subscriber

import com.fasterxml.jackson.databind.ObjectMapper
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import org.springframework.beans.factory.ObjectProvider
import org.springframework.context.SmartLifecycle
import systems.zlink.framework.runtime.host.ZLinkFrameworkRuntime

class OperationalEndpoints(
    private val state: EvidenceStore,
    private val json: ObjectMapper,
    private val endpoint: String,
    private val connections: SubscriberConnections,
    private val observers: FanoutObserverController,
    private val runtime: ObjectProvider<ZLinkFrameworkRuntime>,
) : SmartLifecycle {
    private var server: HttpServer? = null
    private var running = false

    override fun start() {
        if (endpoint.isBlank()) {
            return
        }
        try {
            val uri = URI.create(endpoint)
            val httpServer = HttpServer.create(InetSocketAddress(uri.host, uri.port), 0)
            httpServer.createContext("/health") { exchange ->
                val body = "ok\n".toByteArray(StandardCharsets.UTF_8)
                exchange.sendResponseHeaders(200, body.size.toLong())
                exchange.responseBody.write(body)
                exchange.close()
            }
            httpServer.createContext("/evidence") { exchange ->
                val body = json.writeValueAsBytes(state.snapshot())
                exchange.responseHeaders.add("Content-Type", "application/json")
                exchange.sendResponseHeaders(200, body.size.toLong())
                exchange.responseBody.write(body)
                exchange.close()
            }
            httpServer.createContext("/evidence/wait") { exchange ->
                try {
                    val query = parseQuery(exchange.requestURI.rawQuery.orEmpty())
                    val body = json.writeValueAsBytes(
                        state.waitFor(
                            marker = query["marker"],
                            scenario = query["scenario"],
                            sequence = query["sequence"]?.toIntOrNull(),
                            valueContains = query["contains"],
                            timeoutMillis = query["timeoutMs"]?.toLongOrNull() ?: 30_000L,
                        ),
                    )
                    exchange.responseHeaders.add("Content-Type", "application/json")
                    exchange.sendResponseHeaders(200, body.size.toLong())
                    exchange.responseBody.write(body)
                } catch (error: Exception) {
                    val body = "${error.message ?: error.javaClass.name}\n".toByteArray(StandardCharsets.UTF_8)
                    exchange.sendResponseHeaders(504, body.size.toLong())
                    exchange.responseBody.write(body)
                } finally {
                    exchange.close()
                }
            }
            httpServer.createContext("/status") { exchange ->
                exchange.writeJson(fanoutStatus())
            }
            httpServer.createContext("/connections") { exchange ->
                try {
                    val query = parseQuery(exchange.requestURI.rawQuery.orEmpty())
                    val operation = query["operation"] ?: "list"
                    when (operation) {
                        "connect" -> connections.connect(required(query, "endpoint"))
                        "disconnect" -> connections.disconnect(required(query, "endpoint"))
                        "list" -> Unit
                        else -> throw IllegalArgumentException("unsupported connection operation: $operation")
                    }
                    exchange.writeJson(mapOf("connections" to connections.list()))
                } catch (error: Exception) {
                    exchange.writeText(400, "${error.message ?: error.javaClass.name}\n")
                }
            }
            httpServer.createContext("/observer/start") { exchange ->
                try {
                    val query = parseQuery(exchange.requestURI.rawQuery.orEmpty())
                    observers.start(
                        name = query["name"] ?: "normal",
                        capacity = query["capacity"]?.toIntOrNull() ?: 1,
                        slow = query["slow"]?.toBoolean() ?: false,
                    )
                    exchange.writeJson(mapOf("status" to "started"))
                } catch (error: Exception) {
                    exchange.writeText(400, "${error.message ?: error.javaClass.name}\n")
                }
            }
            httpServer.createContext("/observer/release") { exchange ->
                val query = parseQuery(exchange.requestURI.rawQuery.orEmpty())
                observers.release(query["name"] ?: "slow")
                exchange.writeJson(mapOf("status" to "released"))
            }
            httpServer.createContext("/observer/cancel") { exchange ->
                val query = parseQuery(exchange.requestURI.rawQuery.orEmpty())
                observers.cancel(query["name"] ?: "slow")
                exchange.writeJson(mapOf("status" to "cancelled"))
            }
            httpServer.createContext("/observer/wait") { exchange ->
                try {
                    val query = parseQuery(exchange.requestURI.rawQuery.orEmpty())
                    observers.waitFor(
                        query["name"] ?: "normal",
                        query["timeoutMs"]?.toLongOrNull() ?: 30_000L,
                    )
                    exchange.writeJson(mapOf("status" to "observed"))
                } catch (error: Exception) {
                    exchange.writeText(504, "${error.message ?: error.javaClass.name}\n")
                }
            }
            httpServer.createContext("/observer/evidence") { exchange ->
                exchange.writeJson(observers.snapshot())
            }
            httpServer.start()
            server = httpServer
            running = true
        } catch (error: Exception) {
            throw IllegalStateException("failed to start evidence endpoint $endpoint", error)
        }
    }

    override fun stop() {
        server?.stop(0)
        server = null
        running = false
    }

    override fun isRunning(): Boolean = running

    private fun parseQuery(query: String): Map<String, String> {
        if (query.isBlank()) {
            return emptyMap()
        }
        return query.split("&")
            .mapNotNull { part ->
                val index = part.indexOf("=")
                if (index <= 0) {
                    null
                } else {
                    val key = URLDecoder.decode(part.substring(0, index), StandardCharsets.UTF_8)
                    val value = URLDecoder.decode(part.substring(index + 1), StandardCharsets.UTF_8)
                    key to value
                }
            }
            .toMap()
    }

    private fun fanoutStatus(): Map<String, Any?> {
        val status = runtime.getObject().fanoutRuntime().snapshot(systems.zlink.e2e.kotlin.pubsub.shared.Contracts.EVENT_CHANNEL)
        return mapOf(
            "channelName" to status.channelName(),
            "state" to status.state().name,
            "isReady" to status.isReady,
            "readyPublisherCount" to status.readyPublisherCount(),
            "publishers" to status.publishers().map { publisher ->
                mapOf(
                    "nodeRid" to publisher.nodeRid().toString(),
                    "state" to publisher.state().name,
                    "unavailableReason" to publisher.unavailableReason().map { it.name }.orElse(null),
                )
            },
            "sequence" to status.sequence(),
            "observedAt" to status.observedAt().toString(),
        )
    }

    private fun required(query: Map<String, String>, name: String): String =
        query[name]?.takeIf { it.isNotBlank() } ?: throw IllegalArgumentException("$name is required")

    private fun com.sun.net.httpserver.HttpExchange.writeJson(value: Any) {
        val body = json.writeValueAsBytes(value)
        responseHeaders.add("Content-Type", "application/json")
        sendResponseHeaders(200, body.size.toLong())
        responseBody.use { it.write(body) }
    }

    private fun com.sun.net.httpserver.HttpExchange.writeText(status: Int, value: String) {
        val body = value.toByteArray(StandardCharsets.UTF_8)
        sendResponseHeaders(status, body.size.toLong())
        responseBody.use { it.write(body) }
    }
}
